File: core/src/apps/ethereum/verify_message.py
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from trezor.messages import EthereumVerifyMessage, Success


async def verify_message(msg: EthereumVerifyMessage) -> Success:
    from trezor.crypto.curve import secp256k1
    from trezor.crypto.hashlib import sha3_256
    from trezor.messages import Success
    from trezor.ui.layouts import confirm_signverify, show_success
    from trezor.wire import DataError

    from apps.common.signverify import decode_message

    from .helpers import address_from_bytes, bytes_from_address
    from .sign_message import message_digest

    digest = message_digest(msg.message)
    if len(msg.signature) != 65:
        raise DataError("Invalid signature")
    sig = bytearray([msg.signature[64]]) + msg.signature[:64]

    pubkey = secp256k1.verify_recover(sig, digest)

    if not pubkey:
        raise DataError("Invalid signature")

    pkh = sha3_256(pubkey[1:], keccak=True).digest()[-20:]

    address_bytes = bytes_from_address(msg.address)
    if address_bytes != pkh:
        raise DataError("Invalid signature")

    address = address_from_bytes(address_bytes)

    await confirm_signverify(decode_message(msg.message), address, verify=True)

    await show_success("verify_message", "The signature is valid.")
    return Success(message="Message verified")

File: sdk/apps/ethereum/src/verify_message.rs
use crate::{
    common::decode_message,
    helpers::{address_from_bytes, bytes_from_address},
    layout::confirm_signverify,
    proto::{
        common::{Success, button_request::ButtonRequestType},
        ethereum::VerifyMessage,
    },
    sign_message::message_digest,
};
use trezor_app_sdk::{
    Error, Result, ResultExt,
    crypto::{self, EcCurve, HashingAlgorithm, HasherExt},
    ui,
};

pub fn verify_message(msg: VerifyMessage) -> Result<Success> {
    let digest = message_digest(msg.message.as_slice());

    let mut sig: [u8; 65] = msg
        .signature
        .as_slice()
        .try_into()
        .map_err(|_| Error::DataError("Invalid signature"))?;

    sig.rotate_right(1);
    // Ethereum's wire `v` is `27 + recid` (or `31 + recid` for a compressed-key
    // signature, which never applies here since we only ever recover the
    // uncompressed key) — normalize down to the raw `0..=3` recovery id
    // `ec_recover_pubkey` expects.
    sig[0] = sig[0].wrapping_sub(27) & 3;

    let pubkey = crypto::ec_recover_pubkey(EcCurve::Secp256k1, &sig, &digest)
        .map_err(|_| Error::DataError("Invalid signature"))?;

    let mut hasher = crypto::get_hasher(HashingAlgorithm::Keccak256);
    hasher.update(&pubkey[1..]);
    let pkh_hash = hasher.finalize();

    let address_bytes = bytes_from_address(&msg.address).c()?;

    // We can use unwrap here because the hash is always 32 bytes
    if address_bytes
        != pkh_hash
            .last_chunk::<20>()
            .ok_or(Error::DataError("Hash is too short"))?
    {
        return Err(Error::DataError("Invalid signature"));
    }

    let address = address_from_bytes(&address_bytes, None).c()?;

    confirm_signverify(
        &decode_message(&msg.message).c()?,
        &address,
        true,
        None,
        None,
        false,
    )
    .c()?;

    ui::show_success(ui::ShowSuccess::new(
        tr!("words__title_done"),
        tr!("ethereum__valid_signature"),
        tr!("buttons__continue"),
        None,
        Some("verify_message"),
        ButtonRequestType::Other.into(),
    ))?;

    let msg = Success {
        message: Some("Message verified".into()),
    };

    Ok(msg)
}
